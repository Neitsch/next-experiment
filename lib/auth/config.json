{
  "AUTH0_CLIENT_ID": "_ic7CjcpbO_QPqU-I0NT8C67vk16h9ee",
  "AUTH0_CLIENT_DOMAIN": "condor-club.auth0.com"
}