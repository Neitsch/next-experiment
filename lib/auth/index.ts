import Cookie from "js-cookie";
import jwtDecode from "jwt-decode";

const getQueryParams = () => {
  const params = {};
  window.location.href.replace(
    /([^(?|#)=&]+)(=([^&]*))?/g,
    ($0, $1, $2, $3) => {
      params[$1] = $3;
    },
  );
  return params;
};

export const extractInfoFromHash = () => {
  if (!process.browser) {
    return undefined;
  }
  const { id_token, access_token, state, expires_in } = getQueryParams();
  return { token: id_token, access_token, secret: state, expires_in };
};

export const setToken = (token, accessToken, expiresIn) => {
  if (!process.browser) {
    return;
  }
  const expires = new Date();
  expires.setSeconds(expires.getSeconds() + expiresIn);
  Cookie.set("user", jwtDecode(token), { expires, SameSite: "Lax" });
  Cookie.set("jwt", token, { expires, SameSite: "Lax" });
  Cookie.set("token", accessToken, { expires, SameSite: "Lax" });
};

export const unsetToken = () => {
  if (!process.browser) {
    return;
  }
  Cookie.remove("jwt");
  Cookie.remove("user");
  Cookie.remove("secret");

  // to support logging out from all windows
  window.localStorage.setItem("logout", Date.now());
};

export const getUserFromServerCookie = req => {
  const cookie = getServerCookie(req);
  if (!cookie) {
    return undefined;
  }
  return jwtDecode(cookie);
};

export const getServerCookie = req => {
  if (!req.headers.cookie) {
    return undefined;
  }
  const jwtCookie = req.headers.cookie
    .split(";")
    .find(c => c.trim().startsWith("jwt="));
  if (!jwtCookie) {
    return undefined;
  }
  return jwtCookie.split("=")[1];
};

export const getLocalCookie = () => {
  return Cookie.getJSON("jwt");
};

export const getUserFromLocalCookie = () => {
  return Cookie.getJSON("user");
};

export const setSecret = secret =>
  Cookie.set("secret", secret, { SameSite: "Lax" });

export const checkSecret = secret => {
  return Cookie.get("secret") === secret;
};
