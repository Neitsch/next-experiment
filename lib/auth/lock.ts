import { setSecret } from "./index";

import uuid from "uuid";

const getLock = options => {
  const config = require("./config.json");
  const Auth0Lock = require("auth0-lock").default;
  return new Auth0Lock(
    config.AUTH0_CLIENT_ID,
    config.AUTH0_CLIENT_DOMAIN,
    options,
  );
};

const getBaseUrl = () => `${window.location.protocol}//${window.location.host}`;

const getOptions = container => {
  const secret = uuid.v4();
  setSecret(secret);
  return {
    auth: {
      params: {
        scope: "openid",
        state: secret,
      },
      redirectUrl: `${getBaseUrl()}/auth/signed-in`,
      responseType: "token id_token",
    },
    closable: false,
    container,
  };
};

export const show = container => getLock(getOptions(container)).show();
export const logout = () => getLock({}).logout({ returnTo: getBaseUrl() });
