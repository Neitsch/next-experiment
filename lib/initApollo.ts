import { Hermes } from "apollo-cache-hermes";
import { persistCache } from "apollo-cache-persist";
import { ApolloClient } from "apollo-client";
import { setContext } from "apollo-link-context";
import { HttpLink } from "apollo-link-http";
import { createPersistedQueryLink } from "apollo-link-persisted-queries";
import isomorphicUnfetch from "isomorphic-unfetch";
import localForage from "localforage";

let apolloClient = null;

export const setAuthorizationLink = jwt =>
  setContext(() => ({
    headers: {
      authorization: "Bearer " + jwt,
    },
  }));

export const setupLink = jwt =>
  setAuthorizationLink(jwt).concat(
    createPersistedQueryLink().concat(
      new HttpLink({
        credentials: "same-origin",
        fetch: isomorphicUnfetch,
        uri: "http://localhost:3000/graphql",
      }),
    ),
  );

function create(jwt) {
  const cache = new Hermes({});
  if (process.browser) {
    persistCache({
      cache,
      storage: localForage,
    });
  }
  return new ApolloClient({
    cache,
    connectToDevTools: !!process.browser,
    link: setAuthorizationLink(jwt).concat(
      createPersistedQueryLink().concat(
        new HttpLink({
          credentials: "same-origin",
          fetch: isomorphicUnfetch,
          uri: "http://localhost:3000/graphql",
        }),
      ),
    ),
    ssrMode: !process.browser,
  });
}

export const resetClient = () => {
  apolloClient = null;
};

export default function initApollo(jwt) {
  if (!process.browser) {
    return create(jwt);
  }
  if (!apolloClient) {
    apolloClient = create(jwt);
  }
  return apolloClient;
}
