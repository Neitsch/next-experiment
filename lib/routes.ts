import Routes from "next-routes";

const routes = new Routes().add("index", "/");
export default routes;
export const Link = routes.Link;
export const Router = routes.Router;
