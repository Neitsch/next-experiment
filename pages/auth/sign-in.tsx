import React from "react";

import { show } from "../../lib/auth/lock";

const CONTAINER_ID = "put-lock-here";

class SignIn extends React.Component {
  public componentDidMount() {
    show(CONTAINER_ID);
  }
  public render() {
    return <div id={CONTAINER_ID} />;
  }
}

export default SignIn;
