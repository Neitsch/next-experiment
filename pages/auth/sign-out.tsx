import React from "react";

import { unsetToken } from "../../lib/auth/index";
import { logout } from "../../lib/auth/lock";

export default class SignOut extends React.Component {
  public componentDidMount() {
    unsetToken();
    logout();
  }
  public render() {
    return null;
  }
}
