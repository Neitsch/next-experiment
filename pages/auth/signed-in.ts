/* tslint:disable-next-line */
import Router from "next/router";
import React from "react";

import {
  checkSecret,
  extractInfoFromHash,
  setToken,
} from "../../lib/auth/index";
import initApollo, { setupLink } from "../../lib/initApollo";

export default class SignedIn extends React.Component {
  public componentDidMount() {
    const { token, access_token, secret, expires_in } = extractInfoFromHash();
    if (!checkSecret(secret) || !token) {
      /* tslint:disable-next-line */
      console.error("Something happened with the Sign In request");
      return;
    }
    setToken(token, access_token, parseInt(expires_in));
    const client = initApollo(token);
    client.link = setupLink(token);
    Router.router.push("/");
  }

  public render() {
    return null;
  }
}
