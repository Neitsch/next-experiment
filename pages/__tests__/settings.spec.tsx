import { shallow } from "enzyme";
import React from "react";

import { getLocalCookie, getServerCookie } from "../../lib/auth/index";
import { Router } from "../../lib/routes";
import Settings from "../settings";

jest.unmock("../settings");

describe("Settings page", () => {
  it("Initial props", async () => {
    getLocalCookie.mockReturnValue("local");
    getServerCookie.mockReturnValue("server");
    process.browser = false;
    const retServer = await Settings.getInitialProps({ req: jest.fn() });
    expect(retServer).toEqual({ isAuthenticated: "server" });
    expect(getServerCookie).toHaveBeenCalledTimes(1);
    process.browser = true;
    const retLocal = await Settings.getInitialProps({});
    expect(retLocal).toEqual({ isAuthenticated: "local" });
    expect(getLocalCookie).toHaveBeenCalledTimes(1);
  });
  describe("Renders", () => {
    it("Authenticated", () => {
      const component = shallow(<Settings isAuthenticated={true} />);
      expect(component).toMatchSnapshot();
      expect(Router.pushRoute).not.toHaveBeenCalled();
    });
    it("Unauthenticated", () => {
      const component = shallow(<Settings isAuthenticated={false} />);
      expect(component).toMatchSnapshot();
      expect(Router.pushRoute).toHaveBeenCalled();
    });
  });
});
