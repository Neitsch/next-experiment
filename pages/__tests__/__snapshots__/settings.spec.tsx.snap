// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings page Renders Authenticated 1`] = `<ChangeUsernameComponent />`;

exports[`Settings page Renders Unauthenticated 1`] = `<div />`;
